flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/FlinkMarkChangelogNormaliseProgram.java

```diff
@@ -0,0 +1,150 @@
+                }
+            }
+        }
+        return root;
+    }
+
+    private void gatherTableScanToChangelogNormalizeMap(
+            RelNode curRelNode,
+            Map<TableScan, List<ChangelogNormalizeContext>> map,
+            FlinkRexBuilder rexBuilder) {
+        for (RelNode input : curRelNode.getInputs()) {
+            if (input instanceof StreamPhysicalChangelogNormalize) {
+                StreamPhysicalChangelogNormalize changelogNormalize =
+                        (StreamPhysicalChangelogNormalize) input;
+                if (curRelNode instanceof StreamPhysicalCalc) {
+                    StreamPhysicalCalc calc = (StreamPhysicalCalc) curRelNode;
+                    final List<RexNode> list = getConditions(calc, rexBuilder);
+                    gatherTableScanToChangelogNormalizeMap(
+                            input, ChangelogNormalizeContext.of(changelogNormalize, list), map);
+                }
+            } else {
+                gatherTableScanToChangelogNormalizeMap(input, map, rexBuilder);
+            }
+        }
+    }
+
+    private List<RexNode> getConditions(StreamPhysicalCalc calc, RexBuilder rexBuilder) {
+        final RexProgram program = calc.getProgram();
+        if (program.getCondition() != null) {
+            final RexNode rexNode =
+                    RexUtil.toCnf(
+                            rexBuilder,
+                            FlinkRexUtil.expandSearch(
+                                    rexBuilder, program.expandLocalRef(program.getCondition())));
+            return RelOptUtil.conjunctions(rexNode);
+        }
+        return List.of();
+    }
+
+    private void gatherTableScanToChangelogNormalizeMap(
+            RelNode cur,
+            ChangelogNormalizeContext context,
+            Map<TableScan, List<ChangelogNormalizeContext>> currentMap) {
+        if (cur instanceof TableScan) {
+            currentMap.computeIfAbsent((TableScan) cur, (k) -> new ArrayList<>()).add(context);
+        } else {
+            for (RelNode relNode : cur.getInputs()) {
+                gatherTableScanToChangelogNormalizeMap(relNode, context, currentMap);
+            }
+        }
+    }
+
+    private static class ChangelogNormalizeContext {
+        private final StreamPhysicalChangelogNormalize changelogNormalize;
+        private final List<RexNode> rexNodes;
+
+        public ChangelogNormalizeContext(
+                StreamPhysicalChangelogNormalize changelogNormalize, List<RexNode> rexNodes) {
+            this.changelogNormalize = changelogNormalize;
+            this.rexNodes = rexNodes;
+        }
+
+        public static ChangelogNormalizeContext of(
+                StreamPhysicalChangelogNormalize changelogNormalize, List<RexNode> rexNodes) {
+            return new ChangelogNormalizeContext(changelogNormalize, rexNodes);
+        }
+
+        public StreamPhysicalChangelogNormalize getChangelogNormalize() {
+            return changelogNormalize;
+        }
```

Comment:
Could you unify the access? Either use getters or direct field access.

Reply:
done